### PR TITLE
Regression(ScreenOrientation API) Fullscreen exits right away on Youtube

### PR DESCRIPTION
**Pull request by cdumez**
#### 914c16e8faccc6f14fd3eef8bfaac5065509b98a
<pre>
Regression(ScreenOrientation API) Fullscreen exits right away on Youtube
<a href="https://bugs.webkit.org/show_bug.cgi?id=251396">https://bugs.webkit.org/show_bug.cgi?id=251396</a>
rdar://104269674

Reviewed by Wenson Hsieh.

When entering fullscreen on youtube, the view would lose its window and
then get added to a new Window. This would end up calling
`ScreenOrientationProvider::setWindow()` first with nil then with the new
window.

To determine the screen orientation, the ScreenOrientationProvider currently
needs access to the window object inside platformCurrentOrientation(). If
it didn&apos;t have a window, platformCurrentOrientation() would just return
PortraitPrimary.

When updating the window, setWindow() would call screenOrientationDidChange()
which would call platformCurrentOrientation() again to make sure the cached
orientation is still accurate and to notify the clients that the orientation
has changed otherwise. However, since platformCurrentOrientation() requires
a window, there is no point in calling screenOrientationDidChange() if the
new window is nil. It will just reset the orientation to the PortaitPrimary
value and potentially send an incorrect orientation change notification to
clients. This incorrect notification is what was causing the fullscreen
mode to exit (since fullscreen mode exists on orientation change).

To address this, I made 2 changes:
1. setWindow() now only calls screenOrientationDidChange() if the new
   window is non-nil.
2. platformCurrentOrientation() now returns a std::optional&lt;&gt; to make it
   clear when it wasn&apos;t able to determinate the orientation.
4. In screenOrientationDidChange(), if platformCurrentOrientation() returns
   std::nullopt, we now return early so that we don&apos;t send clients a
   potentially inaccurate orientation. Once the view gets added again to
   a window, we&apos;ll call screenOrientationDidChange() again to make sure
   we send a change notification if needed.

* Source/WebCore/platform/ScreenOrientationProvider.cpp:
(WebCore::ScreenOrientationProvider::screenOrientationDidChange):
(WebCore::ScreenOrientationProvider::currentOrientation):
(WebCore::ScreenOrientationProvider::platformCurrentOrientation):
* Source/WebCore/platform/ScreenOrientationProvider.h:
* Source/WebCore/platform/ios/ScreenOrientationProviderIOS.mm:
(WebCore::ScreenOrientationProvider::platformCurrentOrientation):
(WebCore::ScreenOrientationProvider::setWindow):

Canonical link: <a href="https://commits.webkit.org/259678@main">https://commits.webkit.org/259678@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a490ec94e0e029bac3301b7b4bb1959317a8bbde

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105554 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14643 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38450 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114805 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174959 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109454 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5866 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97848 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12224 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95199 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39705 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94088 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26836 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81408 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7929 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28191 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8314 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4777 "Found 1 new test failure: fast/images/avif-heif-container-as-image.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14053 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47735 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9949 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3575 "Built successfully and passed tests") | | | |

### CODE REVIEW

**Comment by nt1m** [comment]
Is this testable?

**Comment by cdumez** [comment]
> Is this testable?

I need to remove the view from its window and then add it back so I don't think a layout test will work. Maybe an API test though. However, I also need the current orientation to not be PortraitPrimary, that's trickier but maybe an API test can simulate that. I'll give it a shot.

**Comment by cdumez** [comment]
> > Is this testable?
> 
> I need to remove the view from its window and then add it back so I don't think a layout test will work. Maybe an API test though. However, I also need the current orientation to not be PortraitPrimary, that's trickier but maybe an API test can simulate that. I'll give it a shot.

When trying to write an API test, I am running into trouble when trying to change the orientation (since it is portrait by default). I get the following error:
"The current windowing mode does not allow for programmatic changes to interface orientation."

**Comment by hortont424** [comment]
> > Is this testable?
> 
> I need to remove the view from its window and then add it back so I don't think a layout test will work.

UIScriptController:
```
undefined removeViewFromWindow(object callback);
undefined addViewToWindow(object callback);
```

**Comment by cdumez** [comment]
> > > Is this testable?
> > 
> > 
> > I need to remove the view from its window and then add it back so I don't think a layout test will work.
> 
> UIScriptController:
> 
> ```
> undefined removeViewFromWindow(object callback);
> undefined addViewToWindow(object callback);
> ```

Thanks, I'll try that. A layout test would be better than an API test.

File: LayoutTests/fast/dom/Orientation/no-orientation-change-event-when-unparenting-view.html [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                uiController.uiScriptComplete();
+            })
+        });
+    })();`
+}
+
+screen.orientation.addEventListener("change", () => {
+    debug("Orientation changed");
+    orientationChangeCount++;
+    setTimeout(finishJSTest, 10000);
```
Comment (whsieh) [inline LayoutTests/fast/dom/Orientation/no-orientation-change-event-when-unparenting-view.html]: I assume this was just for debugging?

File: LayoutTests/fast/dom/Orientation/no-orientation-change-event-when-unparenting-view.html [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    if (orientationChangeCount == 1) {
+        debug("* Unparenting view, then re-parenting it...");
+        testRunner.runUIScript(getRemoveThenReinsertViewScript(), function(result) { });
+            setTimeout(() => {
+                debug("* No additional orientation change event should have been fired");
+                shouldBe("orientationChangeCount", "1");
+                finishJSTest();
+            }, 100);
```
Comment (whsieh) [inline LayoutTests/fast/dom/Orientation/no-orientation-change-event-when-unparenting-view.html]: Nit - indentation is a bit too far here.

**Review by whsieh — APPROVED** [review]

File: LayoutTests/fast/dom/Orientation/no-orientation-change-event-when-unparenting-view.html (hunk 1 above)
Comment (cdumez) [reply LayoutTests/fast/dom/Orientation/no-orientation-change-event-when-unparenting-view.html]: Oh, yes indeed. Good catch.

**Pull request merged**